fastparquet/api.py

```diff
@@ -782,33 +782,83 @@ def filter_val(op, val, vmin=None, vmax=None):
     -------
     True or False
     """
-    if (op == 'in' and vmax is not None and vmin is not None and
-            vmax == vmin and vmax not in val):
-        return True
+    vmin = _handle_np_array(vmin)
+    vmax = _handle_np_array(vmax)
+    if op == 'in':
+        return filter_in(val, vmin, vmax)
+    if op == 'not in':
+        return filter_not_in(val, vmin, vmax)
     if vmax is not None:
-        if isinstance(vmax, np.ndarray):
-            vmax = vmax[0]
         if op in ['==', '>='] and val > vmax:
             return True
         if op == '>' and val >= vmax:
             return True
-        if op == 'in' and min(val) > vmax:
-            return True
     if vmin is not None:
-        if isinstance(vmin, np.ndarray):
-            vmin = vmin[0]
         if op in ['==', '<='] and val < vmin:
             return True
         if op == '<' and val <= vmin:
             return True
-        if op == 'in' and max(val) < vmin:
-            return True
     if (op == '!=' and vmax is not None and vmin is not None and
             vmax == vmin and val == vmax):
         return True
-    if (op == 'not in' and vmax is not None and vmin is not None and
-            vmax == vmin and vmax in val):
-        return True
 
     # keep this row_group
     return False
+
+
+def _handle_np_array(v):
+    if v is None:
+        return None
+    if isinstance(v, np.ndarray):
+        v = v[0]
+    return v
+
+
+def filter_in(values, vmin=None, vmax=None):
+    """
+    Handles 'in' filters
+
+    op: ['in', 'not in']
+    values: iterable of values
+    vmin, vmax: the range to compare within
+
+    Returns
+    -------
+    True or False
+    """
+    if len(values) == 0:
+        return True
+    if vmax == vmin and vmax is not None and vmax not in values:
+        return True
+    if vmin is None and vmax is None:
+        return False
+
+    sorted_values = sorted(values)
```

Comment:
One option to avoid multiple sorts would be to process all filters once at the top level and pre sort any "in" operators there. This would mean making a copy of the filters.

Reply:
I am sure this is fast enough. If anyone complains, we could reconsider.